packages/asset-swapper/contracts/src/DODOV2Sampler.sol

```diff
@@ -0,0 +1,206 @@
+
+        for (uint256 i = 0; i < numSamples; i++) {
+            uint256 buyAmount = _sampleSellForApproximateBuyFromDODOV2(
+                abi.encode(takerToken, pool, sellBase), // taker token data
+                abi.encode(makerToken, pool, sellBase), // maker token data
+                takerTokenAmounts[i]
+            );
+            // Exit early if the amount is too high for the source to serve
+            if (buyAmount == 0) {
+                break;
+            }
+            makerTokenAmounts[i] = buyAmount;
+        }
+    }
+
+    /// @dev Sample buy quotes from DODO.
+    /// @param registry Address of the registry to look up.
+    /// @param offset offset index for the pool in the registry.
+    /// @param takerToken Address of the taker token (what to sell).
+    /// @param makerToken Address of the maker token (what to buy).
+    /// @param makerTokenAmounts Maker token sell amount for each sample.
+    /// @return sellBase whether the bridge needs to sell the base token
+    /// @return pool the DODO pool address
+    /// @return takerTokenAmounts Taker amounts sold at each maker token
+    ///         amount.
+    function sampleBuysFromDODOV2(
+        address registry,
+        uint256 offset,
+        address takerToken,
+        address makerToken,
+        uint256[] memory makerTokenAmounts
+    )
+        public
+        view
+        returns (bool sellBase, address pool, uint256[] memory takerTokenAmounts)
+    {
+        _assertValidPair(makerToken, takerToken);
+        (pool, sellBase) = _getNextDODOV2Pool(registry, offset, takerToken, makerToken);
+        if (pool == address(0)) {
+            return (sellBase, pool, takerTokenAmounts);
+        }
+        uint256 numSamples = makerTokenAmounts.length;
+        takerTokenAmounts = new uint256[](numSamples);
+
+        takerTokenAmounts = _sampleApproximateBuys(
+            ApproximateBuyQuoteOpts({
+                makerTokenData: abi.encode(makerToken, pool, !sellBase),
+                takerTokenData: abi.encode(takerToken, pool, sellBase),
+                getSellQuoteCallback: _sampleSellForApproximateBuyFromDODOV2
+            }),
+            makerTokenAmounts
+        );
+    }
+
+    function _sampleSellForApproximateBuyFromDODOV2(
+        bytes memory takerTokenData,
+        bytes memory /* makerTokenData */,
+        uint256 sellAmount
+    )
+        private
+        view
+        returns (uint256)
+    {
+        (address takerToken, address pool, bool sellBase) = abi.decode(
+            takerTokenData,
+            (address, address, bool)
+        );
+
+        // We will get called to sell both the taker token and also to sell the maker token
+        // since we use approximate buy for sell and buy functions
+        if (sellBase) {
+            try
+                IDODOV2Pool(pool).querySellBase
+                    { gas: DODO_V2_CALL_GAS }
+                    (address(0), sellAmount)
```

Comment:
Silly question but what sorcery is this? Looks like a fn call without parentheses

Reply:
it's the special solidity syntax to set certain EVM arguments in how the function should be called. Things like gas (give is a gas stipend) and value (how much ETH (wei) to send).

It used to be more of a builder type pattern, until this syntax.

```
IDODOV2Pool(pool).querySellBase.gas(DODO_V2_CALL_GAS).value(MONEY)(address(0), sellAmount)
```